server/server.go

```diff
@@ -59,28 +64,70 @@ func (s *Server) Run(svc transport.Service) error {
 }
 
 func (s *Server) run() error {
-	errChan := make(chan error, 5)
-
+	var g errgroup.Group
 	if s.listeners.mainListener != nil {
-		go func() {
+		g.Go(func() error {
 			err := s.listeners.mainListener.Serve()
-			errChan <- err
-		}()
+			return err
+		})
+	}
+	sigint := make(chan os.Signal, 1)
+	signal.Notify(sigint, os.Interrupt)
+	if s.opts.GracefullFunc != nil {
+		fn := s.opts.GracefullFunc(sigint)
+		g.Go(fn)
+	} else {
+		g.Go(func() error {
+			<-sigint
+			s.Stop()
+			return nil
+		})
 	}
-	go func() {
+	g.Go(func() error {
+		if s.opts.HTTPServer != nil {
+			s.opts.HTTPServer.Handler = s.srv.http
+			err := s.opts.HTTPServer.Serve(s.listeners.HTTP)
+			return err
+		}
 		err := http.Serve(s.listeners.HTTP, s.srv.http)
-		errChan <- err
-	}()
-	go func() {
+		return err
+	})
+	g.Go(func() error {
 		err := s.srv.grpc.Serve(s.listeners.GRPC)
-		errChan <- err
-	}()
+		return err
+	})
 
-	return <-errChan
+	return g.Wait()
 }
 
 // Stop stops the server gracefully.
 func (s *Server) Stop() {
-	// TODO grace HTTP
-	s.srv.grpc.GracefulStop()
+	var g errgroup.Group
+	if s.opts.HTTPServer != nil {
+		g.Go(func() error {
+			return s.opts.HTTPServer.Shutdown(context.Background())
+		})
+	}
+	g.Go(func() error {
+		s.srv.grpc.GracefulStop()
+		return nil
+	})
+	g.Wait()
+}
+
+// StopWithTimeout stops the server gracefully.
+func (s *Server) StopWithTimeout(timeout time.Duration) {
```

Comment:
Func `GracefulStop(ctx)` would look better and more idiomatic.

Reply:
Stop or StopWithTimeout can be used in GracefullFunc to not reimplement it every time.
May be it's more logical to pass context.Context in Stop method? Then pass context in server.Shutdown inside Stop.
May be it's possible to make method Stop With Context. But in this context we can pass in Shutdown only context With Timeout.
That's why here is StopWithTimeout.
What is the reason of Stop method? You can use it outside or inside of Graceful Shutdown according to your situation.
What do you think of it?

Reply:
`Stop` should work as it works now, we don't want to break an API for this. Doing `GracefulStop(context.Background())` in Stop is even worse since we're breaking current behavior without any obvious changes.

I just think that `StopWithTimeout(time.Duration()` should be renamed to `GracefulStop(ctx) ` since it would give user more control and look more in line with stdlib.

Reply:
May be it's better just to add *http.Server variable in server.Server, make *grpc. Server importable and let users to use stdlib funcs?
It's more idiomatic way

Reply:
`clay` aims to unify HTTP and gRPC servers. If you want to tinker with underlying servers/listeners/TCP coons so much that you need a full export - you're better off using the `protoc-gen-goclay` and writing your own server that accepts `transport.Service/ServiceDesc`.